examples_utils/paperspace_utils/metadata_utils.py

```diff
@@ -0,0 +1,206 @@
+    output_dict["Missing Files"] = missing_files
+    if extra_files:
+        logging.warning("Extra files found in local storage: " + str(extra_files))
+    output_dict["Extra files"] = extra_files
+    logging.info("Output dict" + str(output_dict))
+    keys = generated_locally_metadata_files[0].keys()
+    for i in range(len(found_files_metadata)):
+        for key in keys:
+            file_differences = []
+            if found_files_locally[i][key] != found_files_metadata[i][key]:
+                logging.warning(
+                    "Difference in file found and file expected\n"
+                    + "Path: "
+                    + found_files_metadata[i]["path"]
+                    + "\n"
+                    + " Key: "
+                    + key
+                    + "\n"
+                    + " gradient_metadata.json value: "
+                    + str(found_files_metadata[i][key])
+                    + "\n"
+                    + " Local value: "
+                    + str(found_files_locally[i][key])
+                    + "\n"
+                )
+                file_differences.append(
+                    {
+                        "path": found_files_metadata[i]["path"],
+                        "key": key,
+                        "gradient_metadata.json value": str(found_files_metadata[i][key]),
+                        "local value": str(found_files_locally[i][key]),
+                    }
+                )
+        output_dict["file_differences"] = file_differences
+    return output_dict
+
+
+def check_files_match_metadata(dataset_folder: str, compare_hash: bool):
+    """
+    Checks whether files in dataset_folder match the files listed in dataset_folder/METADATA_FILENAME
+
+    Parameters:
+        dataset_folder (str): full or relative path to dataset folder
+        compare_hash (bool): whether or not to compare the md5_hash of the files
+
+    Returns:
+
+
+    """
+    result = {}
+    dataset_folder = Path(dataset_folder)
+    file_list = sorted(list(f for f in dataset_folder.rglob("*") if f.is_file() and f.name != METADATA_FILENAME))
+    gradient_file_arguments = preprocess_list_of_files(dataset_folder, file_list)
+    file_metadata = get_files_metadata(gradient_file_arguments, compare_hash)
+
+    if os.path.isfile(dataset_folder / METADATA_FILENAME):
+        data = json.loads((dataset_folder / METADATA_FILENAME).read_text())
+        result = compare_file_lists(data["files"], file_metadata)
+    else:
+        logging.warning("No metadata file found, no check performed")
+        result = {}
+    return result
+
+
+# Copied from paperspace_automation
+def create_metadata_file(dictionary: dict, path: Path) -> str:
+    content = json.dumps(dictionary, indent=4)
+    file_name = path / METADATA_FILENAME
+    Path(file_name).write_text(content)
+    with open(file_name, "w") as outfile:
+        outfile.write(content)
+    return file_name
+
+
+def get_metadata_file_data(name: str, path: str):
```

Comment:
The name of this function suggests that it will return something, but it doesn't